File: src/pending.rs
use crate::*;

pub(crate) fn apply_pending_audit(
    run: &mut PendingCleanupRun,
    people_count: u32,
    note: Option<String>,
) {
    let audit = AuditEvent {
        at: Local::now(),
        people_count,
        note,
    };
    if run.start_audit.is_none() {
        run.start_audit = Some(people_count);
        run.state = PendingCleanupState::Audited;
    } else if matches!(run.state, PendingCleanupState::NeedsReaudit) {
        run.state = PendingCleanupState::Withdrawing;
    }
    run.latest_audit = Some(people_count);
    run.audits.push(audit);
    run.mark_updated();
}

#[derive(Debug, Deserialize)]
pub(crate) struct PendingCapture {
    #[serde(rename = "capturedAt")]
    pub(crate) captured_at: Option<String>,
    pub(crate) rows: Vec<PendingCaptureRow>,
}

impl PendingCapture {
    pub(crate) fn from_path(path: &PathBuf) -> Result<Self> {
        let raw = fs::read_to_string(path)
            .with_context(|| format!("reading pending capture {}", path.display()))?;
        serde_json::from_str(&raw)
            .with_context(|| format!("parsing pending capture {}", path.display()))
    }
}

#[derive(Debug, Deserialize)]
pub(crate) struct PendingCaptureRow {
    pub(crate) index: u32,
    pub(crate) name: Option<String>,
    #[serde(rename = "profileUrl")]
    pub(crate) profile_url: Option<String>,
    #[serde(rename = "ageText")]
    pub(crate) age_text: Option<String>,
    #[serde(rename = "ageMonths")]
    pub(crate) age_months: Option<u32>,
    pub(crate) eligible: Option<bool>,
    #[serde(rename = "rowText")]
    pub(crate) row_text: Option<String>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct PendingWithdrawResult {
    pub(crate) candidate: PendingWithdrawCandidate,
    pub(crate) status: String,
    #[serde(default)]
    pub(crate) detail: Option<serde_json::Value>,
}

impl PendingWithdrawResult {
    pub(crate) fn from_path(path: &PathBuf) -> Result<Self> {
        let raw = fs::read_to_string(path)
            .with_context(|| format!("reading withdraw result {}", path.display()))?;
        serde_json::from_str(&raw)
            .with_context(|| format!("parsing withdraw result {}", path.display()))
    }

    pub(crate) fn to_withdraw_status(&self) -> (PendingWithdrawStatus, String) {
        match self.status.as_str() {
            "withdrawn-verified" => (
                PendingWithdrawStatus::Withdrawn,
                "salesnav-pending-withdraw-one verified row removed or count decreased".to_string(),
            ),
            "dry-run-withdrawable" => (
                PendingWithdrawStatus::Skipped,
                "dry run found eligible stale invitation".to_string(),
            ),
            "not-eligible" | "row-not-found" => (
                PendingWithdrawStatus::Skipped,
                format!("salesnav-pending-withdraw-one status {}", self.status),
            ),
            other => (
                PendingWithdrawStatus::Failed,
                format!(
                    "salesnav-pending-withdraw-one status {other}; {}",
                    self.detail
                        .as_ref()
                        .map(|value| value.to_string())
                        .unwrap_or_else(|| "no detail".to_string())
                ),
            ),
        }
    }
}

#[derive(Debug, Deserialize)]
pub(crate) struct PendingWithdrawCandidate {
    pub(crate) name: String,
    #[serde(alias = "profileUrl")]
    pub(crate) profile_url: Option<String>,
    #[serde(rename = "age_text", alias = "ageText")]
    pub(crate) age_text: String,
}

pub(crate) fn import_pending_capture(
    run: &mut PendingCleanupRun,
    capture: PendingCapture,
) -> Result<usize> {
    let mut imported = 0;
    for row in capture.rows {
        let Some(name) = row.name.filter(|name| !name.trim().is_empty()) else {
            continue;
        };
        let age_text = row.age_text.unwrap_or_default();
        let age_months = row.age_months.or_else(|| parse_sent_age_months(&age_text));
        let eligible = row
            .eligible
            .unwrap_or_else(|| age_months.is_some_and(|months| months >= run.threshold_months));
        let observation = PendingCandidateObservation {
            imported_at: Local::now(),
            captured_at: capture.captured_at.clone(),
            index: row.index,
            name,
            profile_url: row.profile_url,
            age_text,
            age_months,
            eligible,
            row_text: row.row_text.unwrap_or_default(),
        };
        let existing_index = run.observations.iter().position(|existing| {
            if let (Some(existing_url), Some(new_url)) =
                (&existing.profile_url, &observation.profile_url)
            {
                existing_url == new_url
            } else {
                existing.name == observation.name && existing.age_text == observation.age_text
            }
        });
        if let Some(index) = existing_index {
            run.observations[index] = observation;
        } else {
            run.observations.push(observation);
            imported += 1;
        }
    }
    Ok(imported)
}

pub(crate) fn parse_sent_age_months(age_text: &str) -> Option<u32> {
    let lower = age_text.to_lowercase();
    if lower.contains("year") {
        let count = first_number(&lower).unwrap_or(1);
        return Some(count.saturating_mul(12));
    }
    if lower.contains("month") {
        return Some(first_number(&lower).unwrap_or(1));
    }
    Some(0).filter(|_| {
        lower.contains("today")
            || lower.contains("minute")
            || lower.contains("hour")
            || lower.contains("day")
            || lower.contains("week")
    })
}

pub(crate) fn first_number(value: &str) -> Option<u32> {
    value
        .split(|character: char| !character.is_ascii_digit())
        .find(|part| !part.is_empty())
        .and_then(|part| part.parse().ok())
}

pub(crate) fn record_pending_withdraw_result(
    run: &mut PendingCleanupRun,
    result: PendingWithdrawResult,
    path: PathBuf,
) -> Result<PendingWithdrawEvent> {
    let (status, note) = result.to_withdraw_status();
    let event = PendingWithdrawEvent {
        at: Local::now(),
        name: result.candidate.name,
        profile_url: result.candidate.profile_url,
        age_text: result.candidate.age_text,
        status,
        note: Some(format!("{}; result={}", note, path.display())),
    };
    if status == PendingWithdrawStatus::Withdrawn
        && run.withdrawals.iter().any(|withdrawal| {
            withdrawal.status == PendingWithdrawStatus::Withdrawn
                && withdrawal.name == event.name
                && withdrawal.profile_url == event.profile_url
        })
    {
        bail!("candidate already recorded as withdrawn: {}", event.name);
    }
    run.withdrawals.push(event.clone());
    if !matches!(
        run.state,
        PendingCleanupState::Done | PendingCleanupState::Blocked
    ) {
        run.state = if run.withdraw_capacity_remaining() == 0 {
            PendingCleanupState::FinalReconcile
        } else {
            PendingCleanupState::Withdrawing
        };
    }
    run.mark_updated();
    Ok(event)
}

pub(crate) fn print_pending_plan(plan: &PendingCleanupPlan) {
    match plan {
        PendingCleanupPlan::CaptureMore { reason } => println!("capture more: {reason}"),
        PendingCleanupPlan::WithdrawCandidate {
            name,
            profile_url,
            age_text,
            withdraw_capacity_remaining,
        } => {
            println!("withdraw next stale invitation: {name}");
            println!("age: {age_text}");
            println!(
                "profile_url: {}",
                profile_url.as_deref().unwrap_or("not captured")
            );
            println!("withdraw capacity remaining: {withdraw_capacity_remaining}");
        }
        PendingCleanupPlan::Reaudit { reason } => println!("re-audit: {reason}"),
        PendingCleanupPlan::FinalAudit => println!("final audit"),
    }
}

pub(crate) fn print_pending_status(run: &PendingCleanupRun) {
    println!("run: {}", run.id);
    println!("date: {}", run.date);
    println!("state: {:?}", run.state);
    println!("threshold months: {}", run.threshold_months);
    println!(
        "withdrawn: {}/{}",
        run.withdrawn_count(),
        run.max_withdrawals
    );
    println!(
        "audit: start {:?}, latest {:?}, delta {:?}",
        run.start_audit,
        run.latest_audit,
        run.audited_delta()
    );
    println!("imported observations: {}", run.observations.len());
}

pub(crate) fn render_pending_report(run: &PendingCleanupRun) -> String {
    let mut lines = Vec::new();
    lines.push(format!("# LinkedIn Pending Cleanup {}", run.date));
    lines.push(String::new());
    lines.push(format!("- Run id: `{}`", run.id));
    lines.push(format!("- State: `{:?}`", run.state));
    lines.push(format!("- Threshold: {} months", run.threshold_months));
    lines.push(format!("- Safety cap: {}", run.max_withdrawals));
    lines.push(format!("- Start audit: {:?}", run.start_audit));
    lines.push(format!("- Final/latest audit: {:?}", run.latest_audit));
    lines.push(format!("- Audited delta: {:?}", run.audited_delta()));
    lines.push(format!("- Withdrawn: {}", run.withdrawn_count()));
    lines.push(format!(
        "- Imported pending observations: {}",
        run.observations.len()
    ));
    lines.push(String::new());
    lines.push("## Withdrawn Names".to_string());
    let names = run
        .withdrawals
        .iter()
        .filter(|event| event.status == PendingWithdrawStatus::Withdrawn)
        .map(|event| format!("{} ({})", event.name, event.age_text))
        .collect::<BTreeSet<_>>();
    if names.is_empty() {
        lines.push("- None recorded".to_string());
    } else {
        for name in names {
            lines.push(format!("- {name}"));
        }
    }
    lines.join("\n")
}

File: internal/app/pending.go
package app

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func ApplyPendingAudit(run *PendingCleanupRun, peopleCount uint32, note *string) {
	audit := AuditEvent{At: time.Now(), PeopleCount: peopleCount, Note: note}
	if run.StartAudit == nil {
		run.StartAudit = &peopleCount
		run.State = PendingCleanupStateAudited
	} else if run.State == PendingCleanupStateNeedsReaudit {
		run.State = PendingCleanupStateWithdrawing
	}
	run.LatestAudit = &peopleCount
	run.Audits = append(run.Audits, audit)
	run.MarkUpdated()
}

type PendingCapture struct {
	CapturedAt *string             `json:"capturedAt"`
	Rows       []PendingCaptureRow `json:"rows"`
}

func LoadPendingCapture(path string) (PendingCapture, error) {
	var capture PendingCapture
	if err := readJSONFile(path, &capture, "reading pending capture "+path, "parsing pending capture "+path); err != nil {
		return PendingCapture{}, err
	}
	if capture.Rows == nil {
		capture.Rows = []PendingCaptureRow{}
	}
	return capture, nil
}

type PendingCaptureRow struct {
	Index      uint32  `json:"index"`
	Name       *string `json:"name"`
	ProfileURL *string `json:"profileUrl"`
	AgeText    *string `json:"ageText"`
	AgeMonths  *uint32 `json:"ageMonths"`
	AgeDays    *uint32 `json:"ageDays"`
	Eligible   *bool   `json:"eligible"`
	RowText    *string `json:"rowText"`
}

type PendingWithdrawResult struct {
	Candidate PendingWithdrawCandidate `json:"candidate"`
	Status    string                   `json:"status"`
	Detail    json.RawMessage          `json:"detail"`
}

func LoadPendingWithdrawResult(path string) (PendingWithdrawResult, error) {
	var result PendingWithdrawResult
	if err := readJSONFile(path, &result, "reading withdraw result "+path, "parsing withdraw result "+path); err != nil {
		return PendingWithdrawResult{}, err
	}
	return result, nil
}

func (r PendingWithdrawResult) ToWithdrawStatus() (PendingWithdrawStatus, string) {
	switch r.Status {
	case "withdrawn-verified":
		return PendingWithdrawStatusWithdrawn, "salesnav-pending-withdraw-one verified row removed or count decreased"
	case "dry-run-withdrawable":
		return PendingWithdrawStatusSkipped, "dry run found eligible stale invitation"
	case "not-eligible", "row-not-found":
		return PendingWithdrawStatusSkipped, "salesnav-pending-withdraw-one status " + r.Status
	default:
		detail := "no detail"
		if len(r.Detail) > 0 && string(r.Detail) != "null" {
			detail = string(r.Detail)
		}
		return PendingWithdrawStatusFailed, fmt.Sprintf("salesnav-pending-withdraw-one status %s; %s", r.Status, detail)
	}
}

type PendingWithdrawCandidate struct {
	Name       string  `json:"name"`
	ProfileURL *string `json:"profileUrl"`
	AgeText    string  `json:"age_text"`
}

func (c *PendingWithdrawCandidate) UnmarshalJSON(data []byte) error {
	type candidate struct {
		Name            string  `json:"name"`
		ProfileURL      *string `json:"profileUrl"`
		ProfileURLSnake *string `json:"profile_url"`
		AgeText         string  `json:"age_text"`
		AgeTextCamel    string  `json:"ageText"`
	}
	var value candidate
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}
	c.Name = value.Name
	c.ProfileURL = value.ProfileURL
	if c.ProfileURL == nil {
		c.ProfileURL = value.ProfileURLSnake
	}
	c.AgeText = value.AgeText
	if c.AgeText == "" {
		c.AgeText = value.AgeTextCamel
	}
	return nil
}

func ImportPendingCapture(run *PendingCleanupRun, capture PendingCapture) (int, error) {
	imported := 0
	for _, row := range capture.Rows {
		if row.Name == nil || strings.TrimSpace(*row.Name) == "" {
			continue
		}
		ageText := ""
		if row.AgeText != nil {
			ageText = *row.AgeText
		}
		ageMonths := row.AgeMonths
		if ageMonths == nil {
			ageMonths = ParseSentAgeMonths(ageText)
		}
		ageDays := row.AgeDays
		if ageDays == nil {
			ageDays = ParseSentAgeDays(ageText)
		}
		eligible := false
		if ageDays != nil && run.ThresholdDays > 0 {
			eligible = *ageDays >= run.ThresholdDays
		} else if ageMonths != nil {
			eligible = *ageMonths >= run.ThresholdMonths
		} else if row.Eligible != nil {
			eligible = *row.Eligible
		}
		rowText := ""
		if row.RowText != nil {
			rowText = *row.RowText
		}
		observation := PendingCandidateObservation{
			ImportedAt: time.Now(),
			CapturedAt: capture.CapturedAt,
			Index:      row.Index,
			Name:       *row.Name,
			ProfileURL: row.ProfileURL,
			AgeText:    ageText,
			AgeMonths:  ageMonths,
			AgeDays:    ageDays,
			Eligible:   eligible,
			RowText:    rowText,
		}
		existingIndex := -1
		for i, existing := range run.Observations {
			if existing.ProfileURL != nil && observation.ProfileURL != nil {
				if *existing.ProfileURL == *observation.ProfileURL {
					existingIndex = i
					break
				}
			} else if existing.Name == observation.Name && existing.AgeText == observation.AgeText {
				existingIndex = i
				break
			}
		}
		if existingIndex >= 0 {
			run.Observations[existingIndex] = observation
		} else {
			run.Observations = append(run.Observations, observation)
			imported++
		}
	}
	return imported, nil
}

func ParseSentAgeMonths(ageText string) *uint32 {
	lower := strings.ToLower(ageText)
	if strings.Contains(lower, "year") {
		count := FirstNumber(lower)
		if count == nil {
			value := uint32(1)
			count = &value
		}
		value := *count * 12
		return &value
	}
	if strings.Contains(lower, "month") {
		count := FirstNumber(lower)
		if count == nil {
			value := uint32(1)
			count = &value
		}
		return count
	}
	if strings.Contains(lower, "today") ||
		strings.Contains(lower, "minute") ||
		strings.Contains(lower, "hour") ||
		strings.Contains(lower, "day") ||
		strings.Contains(lower, "week") {
		value := uint32(0)
		return &value
	}
	return nil
}

func ParseSentAgeDays(ageText string) *uint32 {
	lower := strings.ToLower(ageText)
	if strings.Contains(lower, "today") ||
		strings.Contains(lower, "minute") ||
		strings.Contains(lower, "hour") {
		value := uint32(0)
		return &value
	}
	count := FirstNumber(lower)
	if count == nil {
		value := uint32(1)
		count = &value
	}
	if strings.Contains(lower, "year") {
		value := *count * 365
		return &value
	}
	if strings.Contains(lower, "month") {
		value := *count * 30
		return &value
	}
	if strings.Contains(lower, "week") {
		value := *count * 7
		return &value
	}
	if strings.Contains(lower, "yesterday") {
		value := uint32(1)
		return &value
	}
	if strings.Contains(lower, "day") {
		return count
	}
	return nil
}

func FirstNumber(value string) *uint32 {
	parts := strings.FieldsFunc(value, func(r rune) bool { return r < '0' || r > '9' })
	for _, part := range parts {
		if part == "" {
			continue
		}
		parsed, err := strconv.ParseUint(part, 10, 32)
		if err == nil {
			value := uint32(parsed)
			return &value
		}
	}
	return nil
}

func RecordPendingWithdrawResult(run *PendingCleanupRun, result PendingWithdrawResult, path string) (PendingWithdrawEvent, error) {
	status, statusNote := result.ToWithdrawStatus()
	note := fmt.Sprintf("%s; result=%s", statusNote, path)
	event := PendingWithdrawEvent{
		At:         time.Now(),
		Name:       result.Candidate.Name,
		ProfileURL: result.Candidate.ProfileURL,
		AgeText:    result.Candidate.AgeText,
		Status:     status,
		Note:       &note,
	}
	if status == PendingWithdrawStatusWithdrawn {
		for _, withdrawal := range run.Withdrawals {
			if withdrawal.Status == PendingWithdrawStatusWithdrawn && withdrawal.Name == event.Name {
				if (withdrawal.ProfileURL == nil && event.ProfileURL == nil) ||
					(withdrawal.ProfileURL != nil && event.ProfileURL != nil && *withdrawal.ProfileURL == *event.ProfileURL) {
					return PendingWithdrawEvent{}, fmt.Errorf("candidate already recorded as withdrawn: %s", event.Name)
				}
			}
		}
	}
	run.Withdrawals = append(run.Withdrawals, event)
	if run.State != PendingCleanupStateDone && run.State != PendingCleanupStateBlocked {
		if run.WithdrawCapacityRemaining() == 0 {
			run.State = PendingCleanupStateFinalReconcile
		} else {
			run.State = PendingCleanupStateWithdrawing
		}
	}
	run.MarkUpdated()
	return event, nil
}

func writePendingCandidate(path string, candidate PendingCandidateObservation) error {
	raw, err := json.MarshalIndent(candidate, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(path, raw, 0o644)
}
